### ISSUE TITLE
implement a way to enable experiments for testing new language features

### ISSUE THREAD

**Issue by pq** [dart-lang/linter#1412]
Specifically, I'd like to start on #1411, but am not clear on how to turn on the `"set-literals"` language experiment for testing.

Ergonomically, it would be cool if we could do this inlined in our existing tests.  For example, in `test/rule/prefer_collection_literals.dart`, something like:

```dart
@enable-experiments(['set-literals'])
main() {
  ...
  var s7 = new LinkedHashSet.from(['foo', 'bar', 'baz']); // LINT
}
```

would enable the set-literals experiment for the analysis of `main()`.

Another idea would be a directory somewhere called experiments that contains versions of tests to run with specific experiments enabled.  So:

```
test/
  language_experiments/
    set-literals/
      rule/prefer_collection_literals.dart
```

Finally, something more inline would be OK too.

```dart
group('language experiments', () {
  group('set-literals', () {
    test('prefer_collection_literals', () {
      var testSource = ...;
      var lintRule = ...;
      var results = lintWithExperiments(lintRule, testSource, 'set-literals');
      expect(results, orderedEquals(...));
     });  
  });
});
```

/cc @bwilkerson

**Comment by bwilkerson** [comment]
I'm not familiar enough with linter's testing infrastructure to know how a test like `test/rule/prefer_collection_literals.dart` is run, so I don't know what is and isn't possible. I would hope that those data files are analyzed in such a way that any analysis options files in the path would be found.

If that's true, then using a directory structure such as you suggested seems like the easiest solution. We would just need to add an analysis options file in each directory that corresponds to an experiment (such as `set-literals`) that enables the experiment for all of the files in that directory. Seems like it would also minimize duplication of information.

**Comment by pq** [comment]
> I would hope that those data files are analyzed in such a way that any analysis options files in the path would be found.

👍 That would be great.  To get that working it looks like me may need to thread some things through in `analyzer`.

Here's how we lint each rule.

https://github.com/dart-lang/linter/blob/36316c3995344b2d169b1d08afce26fc47b47d8f/test/rule_test.dart#L338-L343

As it stands there isn't a way to pass analysis options (or an options file) but it seems really reasonable to me to add it.  A natural place might be `LinterOptions`:

https://github.com/dart-lang/sdk/blob/eaeffd269535ee9deb61512e3470a2e34559cc85/pkg/analyzer/lib/src/lint/linter.dart#L292

If that sounds reasonable, let me know and I can put something together.

**Comment by bwilkerson** [comment]
Yes, that's one way to do it, and it seems to be in keeping with the current architecture.

The other option is to build the file structure a user would build (in a `MemoryResourceProvider`) and use `AnalysisContextCollection` to build everything the way it will (eventually) be built by the command-line analyzer. It would be more work in the short term, but ought to avoid needing to extend the API later to pass in more and more options. Still, I'm not sure it's worth doing right now.

**Comment by pq** [comment]
Implemented w/ #1439
